simdev/module/github/gather_popular_repos_stage.py

```diff
@@ -0,0 +1,197 @@
+    def name(self):
+        """
+        :return: name of the stage
+        """
+        return "Gather popular repositories"
+
+    def __init__(
+        self,
+        source_repos: List[str],
+        max_popular_repos_num: int,
+        page_limit: int,
+        api_tokens: List[str],
+        stargazers_fetch_process_count: int,
+    ):
+        """
+        Initialize stage with input params
+        :param source_repos: list of URLs to
+        GitHub repositories
+        :param max_popular_repos_num: max number of produced (most popular) repositories
+        :param page_limit: GitHub API page limit
+        :param api_tokens: list of GitHub API tokens
+        :param stargazers_fetch_process_count the number
+        of processes to fetch stargazers in
+        """
+        self._context = PopularReposContext(
+            source_repos,
+            max_popular_repos_num,
+            page_limit,
+            api_tokens
+        )
+        self._api_wrapper = GithubApiWrapper(api_tokens=self._context.api_tokens)
+        self.stargazers_fetch_process_count = stargazers_fetch_process_count
+
+        # Cache the main fetching function
+        self._count_top_repos = PipelineCache.memory.cache(self._count_top_repos)
+
+    def run(self, pipeline: Pipeline):
+        """
+        First of all, we fetch all the stargazers for requested
+        repositories (merge them not to include duplicates).
+        Then we run over all the stargazers and fetch their
+        starred repositories (saving intermediate results if
+        possible). After that we accumulate all the results and
+        determine top-n repositories starred by initial
+        stargazers.
+        :param pipeline: pipline inside which the stage is run
+        """
+        self._context.popular_repositories = self._count_top_repos()
+
+    def _count_top_repos(self):
+        """
+        Firstly, we retrieve all the stargazers for the requested repositories,
+        merging them to avoid duplicates.
+        Next, we iterate through each stargazer and fetch their
+        starred repositories, saving intermediate results
+        whenever feasible.
+        Then, we gather all the results and identify the top-n
+        repositories that have been starred
+        by the original stargazers
+        :return: The list of top-n repositories, ranked by
+        the number of stars received from the initial stargazers
+        """
+        source_tqdm = tqdm(
+            self._context.source_repositories,
+            desc="Coursing through source repositories",
+        )
+        repo_star_counter = Counter()
+
+        # Acknowledged stargazers maintained to avoid duplicates between several
+        # source repositories
+        acked_stargazers: Set[str] = set()
+
+        for source_repo in source_tqdm:
+            stargazers_context = StargazersContext(repo=source_repo)
+            stargazers_context.stargazers = self._api_wrapper.fetch_stargazers(
```

Comment:
I would recommend to cache `fetch_stargazers` (if it's not yet cached - haven't gone through all PR yet) - it will save API calls to github API

Reply:
it's cached - good